sdk/go/pulumi/rpc.go

```diff
@@ -308,90 +326,101 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
+		}
+		return sv, true, nil
 	case v.IsArray():
 		arr := v.ArrayValue()
 		rv := make([]interface{}, len(arr))
+		secret := false
 		for i, e := range arr {
-			ev, err := unmarshalPropertyValue(e)
+			ev, esecret, err := unmarshalPropertyValue(e)
+			secret = secret || esecret
 			if err != nil {
-				return nil, err
+				return nil, false, err
 			}
 			rv[i] = ev
 		}
-		return rv, nil
+		return rv, secret, nil
 	case v.IsObject():
 		m := make(map[string]interface{})
+		secret := false
 		for k, e := range v.ObjectValue() {
-			ev, err := unmarshalPropertyValue(e)
+			ev, esecret, err := unmarshalPropertyValue(e)
+			secret = secret || esecret
 			if err != nil {
-				return nil, err
+				return nil, false, err
 			}
 			m[string(k)] = ev
 		}
-		return m, nil
+		return m, secret, nil
 	case v.IsAsset():
 		asset := v.AssetValue()
 		switch {
 		case asset.IsPath():
-			return NewFileAsset(asset.Path), nil
+			return NewFileAsset(asset.Path), false, nil
 		case asset.IsText():
-			return NewStringAsset(asset.Text), nil
+			return NewStringAsset(asset.Text), false, nil
 		case asset.IsURI():
-			return NewRemoteAsset(asset.URI), nil
+			return NewRemoteAsset(asset.URI), false, nil
 		}
-		return nil, errors.New("expected asset to be one of File, String, or Remote; got none")
+		return nil, false, errors.New("expected asset to be one of File, String, or Remote; got none")
 	case v.IsArchive():
 		archive := v.ArchiveValue()
+		secret := false
 		switch {
 		case archive.IsAssets():
 			as := make(map[string]interface{})
 			for k, v := range archive.Assets {
-				a, err := unmarshalPropertyValue(resource.NewPropertyValue(v))
+				a, asecret, err := unmarshalPropertyValue(resource.NewPropertyValue(v))
+				secret = secret || asecret
 				if err != nil {
-					return nil, err
+					return nil, false, err
 				}
 				as[k] = a
 			}
-			return NewAssetArchive(as), nil
+			return NewAssetArchive(as), secret, nil
 		case archive.IsPath():
-			return NewFileArchive(archive.Path), nil
+			return NewFileArchive(archive.Path), secret, nil
 		case archive.IsURI():
-			return NewRemoteArchive(archive.URI), nil
+			return NewRemoteArchive(archive.URI), secret, nil
 		default:
 		}
-		return nil, errors.New("expected asset to be one of File, String, or Remote; got none")
+		return nil, false, errors.New("expected asset to be one of File, String, or Remote; got none")
 	default:
-		return v.V, nil
+		return v.V, false, nil
 	}
 }
 
 // unmarshalOutput unmarshals a single output variable into its runtime representation.
-func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) error {
+// returning a bool that indicates secretness
+func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) (bool, error) {
 	contract.Assert(dest.CanSet())
 
 	// Check for nils and unknowns. The destination will be left with the zero value.
 	if v.IsNull() || v.IsComputed() || v.IsOutput() {
-		return nil
+		return v.IsSecret(), nil
```

Comment:
This will always be `false`.

```suggestion
		return false, nil
```